atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ClusterMetadataUtils.java

```diff
@@ -0,0 +1,190 @@
+     * @param keyspace The keyspace containing the partition keys we want to map
+     * @param partitionKeyTokens The Cassandra tokens for the partition keys we want to map
+     * @return Mapping of Node to token ranges its host contains, where every partition key in the specified
+     * list is present in one of the token ranges on one host
+     */
+    @SuppressWarnings("ReverseDnsLookup")
+    public static Map<InetSocketAddress, Set<TokenRange>> getTokenMapping(
+            Collection<InetSocketAddress> nodeSet, Metadata metadata, String keyspace, Set<Token> partitionKeyTokens) {
+
+        Map<String, InetSocketAddress> nodeMetadataHostMap = StreamEx.of(nodeSet)
+                .mapToEntry(InetSocketAddress::getHostName)
+                .invert()
+                .toMap();
+        Map<Host, List<TokenRange>> hostToTokenRangeMap =
+                getTokenMappingForPartitionKeys(metadata, keyspace, partitionKeyTokens);
+
+        return EntryStream.of(hostToTokenRangeMap)
+                .mapValues(list -> (Set<TokenRange>) new HashSet<>(list))
+                .mapKeys(host -> {
+                    String hostname = host.getEndPoint().resolve().getHostName();
+                    Preconditions.checkArgument(
+                            nodeMetadataHostMap.containsKey(hostname),
+                            "Did not find corresponding Node to run repair",
+                            SafeArg.of("hostname", hostname));
+                    return nodeMetadataHostMap.get(hostname);
+                })
+                .toMap();
+    }
+
+    private static Map<Host, List<TokenRange>> getTokenMappingForPartitionKeys(
+            Metadata metadata, String keyspace, Set<Token> partitionKeyTokens) {
+        Set<TokenRange> tokenRanges = metadata.getTokenRanges();
+        SortedMap<Token, TokenRange> tokenRangesByEnd =
+                StreamEx.of(tokenRanges).mapToEntry(TokenRange::getEnd).invert().toSortedMap();
+        Set<TokenRange> ranges = getSmallTokenRangeForKey(metadata, partitionKeyTokens, tokenRangesByEnd);
+        Multimap<Host, TokenRange> tokenMapping = ArrayListMultimap.create();
+        ranges.forEach(range -> {
+            List<Host> hosts = ImmutableList.copyOf(metadata.getReplicas(quotedKeyspace(keyspace), range));
+            if (hosts.isEmpty()) {
+                throw new SafeIllegalStateException(
+                        "Failed to find any replicas of token range for repair",
+                        SafeArg.of("tokenRange", range.toString()),
+                        SafeArg.of("keyspace", quotedKeyspace(keyspace)));
+            }
+            hosts.forEach(host -> tokenMapping.put(host, range));
+        });
+        return KeyedStream.stream(tokenMapping.asMap())
+                .map(trs -> (List<TokenRange>) new ArrayList<>(trs))
+                .collectToMap();
+    }
+
+    // TODO(gs): copy over tests?
+    private static Set<TokenRange> getSmallTokenRangeForKey(
```

Comment:
This method can do with some nice refactoring. I would rename it to `getMinimalSetOfRangesForTokens`